### PR TITLE
supersedes [TEST] I/O module snippets

### LINKED ISSUE #769: Structure file bugs
While working on the io snippets #672 I encountered two bugs in structure file output:

1. Trying to write a structured_seq object causes an compiler error (same as #671 )
2. The `push_pack` overload for records was missing so outputting records lead to incorrect ordering. 

Both issues will be fixed in the snippets PR

### PR DESCRIPTION
**Pull request by smehringer**
supersedes #672 

Resolves #769

### CODE REVIEW

**Comment by smehringer** [comment]
@h-2 Everything works now. There is only the alignment_file snippet that relies on alignment file input

File: test/snippet/io/alignment_file/alignment_file_output.cpp [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+{
+//! [set_header]
+alignment_file_output fout{tmp_dir/"my.sam"};
+
+// add information to the header of the file.
+fout.header().comments.push_back("This is a comment");
+//! [set_header]
+}
+
+// \todo TODO: uncomment once seqan3::alignment_file_input is implemented.
```
Comment (h-2) [inline test/snippet/io/alignment_file/alignment_file_output.cpp]: This really has the potential to be forgotten...

File: test/snippet/io/alignment_file/alignment_file_output.cpp [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        { "GATA"_dna5, "Third" }
+    }; // a range of "records"
+
+    fout = range; // will iterate over the records and write them
+}
+
+{
+//! [input_range]
+// file format conversion in one line:
+alignment_file_output{"/tmp/output.sam"} = sequence_file_input{"/tmp/input.fasta"};
```
Comment (h-2) [inline test/snippet/io/alignment_file/alignment_file_output.cpp]: Shouldn't this example be 
```cpp
alignment_file_output{"/tmp/output.sam"} = alignment_file_input{"/tmp/input.bam"};
```

i.e. conversion between two formats of the same file type. I mean in the above example, it actually works to transforma sequence_file to an alignment_file, but it doesn't really make a lot of sense.

Also you still use `/tmp` here and not a real temporary directory.

File: test/snippet/io/sequence_file/sequence_file_input.cpp [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+fout.emplace_back("GGAGTATAATATATATATATATAT"_dna4, "Test3");
+}
+
+{
+// Create a /tmp/my_seq_qual.fasta file.
+sequence_file_output fout{tmp_dir/"my.fastq", fields<field::ID, field::SEQ_QUAL>{}};
+std::vector<qualified<dna4, phred42>> my_qualified{qualified<dna4, phred42>{'A'_dna4, phred42{7}},
+                                                   qualified<dna4, phred42>{'C'_dna4, phred42{6}},
+                                                   qualified<dna4, phred42>{'G'_dna4, phred42{5}},
+                                                   qualified<dna4, phred42>{'T'_dna4, phred42{4}}};
```
Comment (h-2) [inline test/snippet/io/sequence_file/sequence_file_input.cpp]: wouldn't this also work (an be shorter):
```cpp
std::vector<qualified<dna4, phred42>> my_qualified{{'A'_dna4, phred42{7}},
                                                   {'C'_dna4, phred42{6}},
                                                   {'G'_dna4, phred42{5}},
                                                   {'T'_dna4, phred42{4}}};
```

Also the phreds have literals now ☝️

File: test/snippet/io/structure_file/structure_file_output.cpp [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+//
+// for (auto & r : fin)
+// {
+//     if (criteria)// r fulfills some criterium
+//     {
+//         fout.push_back(r);
+//     }
+// }
+// //! [pass_rec]
+// }
```
Comment (h-2) [inline test/snippet/io/structure_file/structure_file_output.cpp]: whats this?

File: test/snippet/io/alignment_file/alignment_file_output.cpp (hunk 2 above)
Comment (smehringer) [reply test/snippet/io/alignment_file/alignment_file_output.cpp]: well maybe FASTA -> SAm does not make sense but SAM -> FASTA actually does. There is even a samtools command especially for this. But of course coversion within the same type makes more sense in the first place... we just don't have another format :trollface:

File: test/snippet/io/alignment_file/alignment_file_output.cpp (hunk 1 above)
Comment (smehringer) [reply test/snippet/io/alignment_file/alignment_file_output.cpp]: I could create an issue for it...?

File: test/snippet/io/alignment_file/alignment_file_output.cpp (hunk 2 above)
Comment (h-2) [reply test/snippet/io/alignment_file/alignment_file_output.cpp]: > SAM -> FASTA actually does. There is even a samtools command especially for this. 

Yes, but that's why we have the sequence_file_format_sam in addition to te alignment_file_format_sam, or not?

File: test/snippet/io/alignment_file/alignment_file_output.cpp (hunk 1 above)
Comment (h-2) [reply test/snippet/io/alignment_file/alignment_file_output.cpp]: Or leave it commented out so its visible in the docs? Or add \todo to docs?

File: test/snippet/io/structure_file/structure_file_output.cpp (hunk 4 above)
Comment (smehringer) [reply test/snippet/io/structure_file/structure_file_output.cpp]: This just caught a fat bug.. Structure file was missing the push_back overload for records (not mere tuples) so the order fields was not correctly parsed.

**Comment by smehringer** [comment]
@h-2 Can you take another look?

**Comment by h-2** [comment]
Can you stash your last commits?

**Comment by smehringer** [comment]
sure I justed wanted to do that after you reviewd

**Comment by smehringer** [comment]
Can we merge this or shall I request another review?

**Pull request merged**